lib/rubygems/test_case.rb

```diff
@@ -1623,10 +1500,3 @@ def self.key_path(key_name)
 end
 
 require 'rubygems/test_utilities'
-tmpdirs = []
-tmpdirs << (ENV['GEM_HOME'] = Dir.mktmpdir("home"))
-tmpdirs << (ENV['GEM_PATH'] = Dir.mktmpdir("path"))
-pid = $$
-END {tmpdirs.each {|dir| Dir.rmdir(dir)} if $$ == pid}
-Gem.clear_paths
-Gem.loaded_specs.clear
```

Comment:
Most of the diff in this PR is actually part of #2742, it shows up here because I didn't have time to rebase after #2742 got merged.